<img width="350" alt="logo" src="https://github.com/AndergassenLab/Allelome.LINK/assets/146940261/587779dd-a554-4c99-b6b9-a2b60e3e1297" align="right">
# Resource generated by *Hasenbein T., Andergassen, D. et al.*
To interactively access the resource generated by *Hasenbein T., Andergassen, D. et al.*, use JSON files for mice and human samples from the resource folder. 
Candidate predictions can then be viewed via the Integrative Genomics Viewer (IGV) at [https://igv.org/app/](https://igv.org/app/). 

Publication: *Hasenbein T., Andergassen, D. et al.*

IGV: *James T. Robinson, Helga Thorvaldsdóttir, Wendy Winckler, Mitchell Guttman, Eric S. Lander, Gad Getz, Jill P. Mesirov. Integrative Genomics Viewer. Nature Biotechnology 29, 24–26 (2011). A public access version is also available: PMC3346182.*

Please load JSON files as a session. Upon loading the file of interest, you can see the following output:

## 5.1 Example mice: Heart 
<img width="700" alt="mice_example" src="https://github.com/AndergassenLab/Allelome.LINK/assets/146940261/d79d4879-a786-48ea-8294-1dc6f06e08fb">


| Track      | Description   | 
| ------------- | ------------- | 
| Allelome.LINK_Ki_9w_enhancing| Green arcs show enhancing linkages, where a ncRNA was predicted to have an enhancing role on the connected protein-coding target gene. By clicking on the linkage, both interaction loci are named, and the linkage score is displayed | 
| Merged_fwd | RNA-seq track as a representative example. The track was obtained from a single replicate out of the three replicates where the linkages are based on. Reads are shown for the forward strand only and split towards the maternal (red) and paternal (blue) allele | 
| Allelome.PRO_Ki_9w | Allelome.PRO v2.0 output naming the locus, the median number of reads, the min. allelic score and the median allelic ratio. The color denotes the allelic bias (red: maternal, green: biallelic, blue: paternal)| 
| Refseq Genes | RefSeq gene annotation | 
| Merged_rev | RNA-seq track as a representative example. The track was obtained from a single replicate out of the three replicates where the linkages are based on. Reads are shown for the reverse strand only and split towards the maternal (red) and paternal (blue) allele | 
| Allelome.LINK_Ki_9w_repressive| Red arcs show repressive linkages, where a ncRNA was predicted to have a repressive role on the connected protein-coding target gene. By clicking on the linkage, both interaction loci are named, and the linkage score is displayed  |

## 5.2 Example human: Heart - Left Ventricle
<img width="700" alt="mice_example" src="https://github.com/AndergassenLab/Allelome.LINK/assets/146940261/cf728db9-3bbd-46ef-99a1-9dbce639c9a2">

| Track      | Description   | 
| ------------- | ------------- | 
| Cells-Cultured-fibroblasts_enh| Green arcs show enhancing linkages, where a ncRNA was predicted to have an enhancing role on the connected protein-coding target gene. By clicking a linkage, both interaction loci are named, along with the linkage score and the GTEx sample where the linkage was found. Different arcs represent different samples. | 
| RNAseq.bw | RNA-seq track as a representative example. The track was obtained from a single sample of the tissue on which the linkages are based.| 
| annotation.bed | GENCODE v26 GRCh38 annotation | 
| Cells-Cultured-fibroblasts_rep| Red arcs show repressive linkages, where a ncRNA was predicted to have a repressive role on the connected protein-coding target gene. By clicking on the linkage, both interaction loci are named, along with the linkage score and the GTEx sample where the linkage was found. Different arcs represent different sample.  |
| gwas_snps.bed | GWAS SNPs derived from the Ebi GWAS catalog | 
